File: libs/fs/CoreCrypto/CoreCiphers.fs
(* Copyright (C) 2012--2015 Microsoft Research and INRIA *)

module CoreCiphers
open Bytes

open CryptoProvider
open Org.BouncyCastle.Crypto.Engines
open Org.BouncyCastle.Crypto.Parameters

type key   = bytes
type iv    = bytes
type adata = bytes

let encrypt cipher omode key plain =
    let engine = CoreCrypto.BlockCipher ForEncryption cipher omode (cbytes key) in
        abytes (engine.Process (cbytes plain))

let decrypt cipher omode key encrypted =
    let engine = CoreCrypto.BlockCipher ForDecryption cipher omode (cbytes key) in
        abytes (engine.Process (cbytes encrypted))

let aes_cbc_encrypt key iv plain     = encrypt cipher.AES (Some (CBC (cbytes iv))) key plain
let aes_cbc_decrypt key iv encrypted = decrypt cipher.AES (Some (CBC (cbytes iv))) key encrypted

let des3_cbc_encrypt key iv plain     = encrypt DES3 (Some (CBC (cbytes iv))) key plain
let des3_cbc_decrypt key iv encrypted = decrypt DES3 (Some (CBC (cbytes iv))) key encrypted

let aes_gcm_encrypt key iv ad plain =
    let acipher = acipher.AES
    let amode   = GCM (cbytes iv, cbytes ad)
    let engine  = CoreCrypto.AeadCipher ForEncryption acipher amode (cbytes key) in
        abytes (engine.Process (cbytes plain))

let aes_gcm_decrypt key (iv : iv) (ad : adata) plain =
    let acipher = acipher.AES
    let amode   = GCM (cbytes iv, cbytes ad)
    let engine  = CoreCrypto.AeadCipher ForDecryption acipher amode (cbytes key) in
        try
            Some (abytes (engine.Process (cbytes plain)))
        with AEADFailure -> None

type rc4engine = RC4Engine of StreamCipher

let rc4create (key : key) =
    let engine = CoreCrypto.StreamCipher ForEncryption (* ignored *) RC4 (cbytes key) in
        RC4Engine engine

let rc4process (RC4Engine engine) (input : bytes) =
    abytes (engine.Process (cbytes input))

File: libs/fst/CoreCrypto_DEPRECATED/CoreCiphers.ml
(* -------------------------------------------------------------------- *)
open Bytes
open CoreCrypto

(* -------------------------------------------------------------------- *)
type key = bytes
type iv  = bytes

(* -------------------------------------------------------------------- *)
let encrypt cipher omode key plain =
  let engine = blockcipher ForEncryption cipher omode (cbytes key) in
  abytes (engine.bc_process (cbytes plain))

(* -------------------------------------------------------------------- *)
let decrypt cipher omode key encrypted =
  let engine = CoreCrypto.blockcipher ForDecryption cipher omode (cbytes key) in
  abytes (engine.bc_process (cbytes encrypted))

(* -------------------------------------------------------------------- *)
let aes_cbc_encrypt key iv plain     = encrypt AES (Some (CBC (cbytes iv))) key plain
let aes_cbc_decrypt key iv encrypted = decrypt AES (Some (CBC (cbytes iv))) key encrypted

(* -------------------------------------------------------------------- *)
let des3_cbc_encrypt key iv plain     = encrypt DES3 (Some (CBC (cbytes iv))) key plain
let des3_cbc_decrypt key iv encrypted = decrypt DES3 (Some (CBC (cbytes iv))) key encrypted

(* -------------------------------------------------------------------- *)
type rc4engine = RC4Engine of streamcipher

let rc4create (key : key) =
  RC4Engine (CoreCrypto.streamcipher ForEncryption RC4 (cbytes key))

let rc4process (RC4Engine engine) (input : bytes) =
  abytes (engine.sc_process (cbytes input))

let aes_gcm_encrypt (key:key) (iv:iv) (ad:bytes) (plain:bytes) = failwith "unimplemented"
let aes_gcm_decrypt (key:key) (iv:iv) (ad:bytes) (plain:bytes) = failwith "unimplemented"